File: year2018/day02/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  twice, thrice = 0, 0
  for row in input:
    counter = {}

    for c in row:
      counter[c] = counter.get(c, 0) + 1

    twice += 1 if 2 in counter.values() else 0
    thrice += 1 if 3 in counter.values() else 0

  return twice * thrice


def part_2():
  for x in range(len(input)):
    for y in range(len(input)):
      if x != y and len(input[x]) == len(input[y]):
        difference, position = 0, 0

        for z in range(len(input[x])):
          if input[x][z] != input[y][z]:
            position, difference = z, difference + 1

        if difference == 1:
          return (input[x][:position] + input[x][position + 1:]).rstrip()


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [row for row in file]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2018/day02/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static String[] IDS;

    public static void main(String[] args) { 
        try {
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        int twice = 0;
        int thrice = 0;
        for (String row : IDS) {
            Map<Character, Integer> counter = new HashMap<>();
            for (char c : row.toCharArray()) {
                counter.put(c, counter.getOrDefault(c, 0) + 1);
            }
            twice += counter.values().contains(2) ? 1 : 0;
            thrice += counter.values().contains(3) ? 1 : 0;
        }
        return twice * thrice;
    }
    
    public static String part2() {
        for (int x = 0; x < IDS.length; x++) {
            for (int y = 0; y < IDS.length; y++) {
                int difference = 0;
                int position = 0;
                for (int z = 0; z < IDS[x].length(); z++) {
                    if (IDS[x].charAt(z) != IDS[y].charAt(z)) {
                        position = z;
                        difference++;
                    }
                }
                if (difference == 1) {
                    return IDS[x].substring(0, position) + IDS[x].substring(position + 1);
                }
            }
        }
        return null;
    }

    public static void setup(String input) {
        IDS = input.split("\n");
    }
}